protocol_connect.go

```diff
@@ -0,0 +1,963 @@
+// Copyright 2021-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connect
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+
+	errorv1 "github.com/bufbuild/connect-go/internal/gen/go/connect/error/v1"
+)
+
+const (
+	connectUnaryHeaderCompression           = "Content-Encoding"
+	connectUnaryHeaderAcceptCompression     = "Accept-Encoding"
+	connectUnaryTrailerPrefix               = "Trailer-"
+	connectStreamingHeaderCompression       = "Connect-Encoding"
```

Comment:
Maybe we can converge content-encoding for both unary and streaming since they are using the same key value?